src/python/WMCore/Services/Rucio/Rucio.py

```diff
@@ -478,6 +480,44 @@ def createReplicationRule(self, names, rseExpression, scope='cms', copies=1, **k
         except AccessDenied as ex:
             msg = "AccessDenied creating DID replication rule. Error: %s" % str(ex)
             raise WMRucioException(msg)
+        except DuplicateRule as ex:
+            # NOTE:
+            #    This exception will be thrown by Rucio even if a single Did has
+            #    a duplicate rule. In this case all the rest of the Dids will be
+            #    ignored, which in general should be addressed by Rucio. But since
+            #    it is not, we should break the list of Dids and proceed one by one
+            # NOTE:
+            #    This thing here may be slow, because it will wait for Rucio to
+            #    return the history of rules per every Did, but no shorter path exists
+            msg = "A duplicate rule for: \naccount: %s \ndids: %s \nrseExpression: %s\ncopies: %s\n"
+            ruleHistory = []
+            for did in dids:
```

Comment:
This we quickly discussed over slack. We should have a separate method for dealing with the rules history.
In addition to that, can you please add a comment to what a duplicate rule is? AFAIK, it's considered as a unique key of:
rucio account + did + rse expression + scope

If this is correct, then I don't think we need to break down the list of dids and fetch the history for every single of them. We would either have a duplicate rule (for all of them), or for none of them.
@nsmith- might be able to clarify it though.

Reply:
Yes as discussed earlier I am working on that. 

And yes, the unique constrain you mention is correct with just one minor but really important detail. We do not submit per single `did` but rather a whole list of `dids`. And this constrain looked like an Oracle constrained to me (one of the exceptions I received was kind of pointing in that direction), and since it seems to be a general `unique` constrain it is raising the `DuplicateRule` exception regardless of the number of the `dids` in the submission request violating it. Which immediately skips the rule creation for the ones that are in the same list but are not violating this requirement. and masks that quite well.  Which immediately turns into a silent bug. So Yes, I am quite sure we must break the list in pieces here. I tested  this  several times in many different configurations.My oppinion is that  this think should be addressed in Rucio though, because asking the full history per every single did from our side may result in prolonged queries.

Reply:
indeed the tuple `rucio account + did (=name, scope) + rse expression` is the unique key for rules

Reply:
Yes, but the `did` is still single not plural :) and we are in the situation I explained above. Could this ever be addressed in Rucio or it would be too complicate and we just keep it as it is. The way I handle this is doing the job anyway. We are just risking a little bit of longish queries.

Reply:
Ok, I see. If you provide multiple DIDs, it just loops over them and makes a rule for each. It may be that, given one DID in the list hits the duplicate rule exception, the others are processed as normal, but I am not sure.

Reply:
Todor, please correct me if I misunderstood your comment.
When a list of dids is passed to the Rucio server, Rucio does NOT deal with them in a single transaction? Meaning that some dids might succeed to have a rule created, while others not. Have you actually seen this behaviour in your tests?
If that's correct, then that's a serious problem that needs to be addressed on the server side. Otherwise clients have to implement all these extra checks in order to get their multi-dids rules going through.

The rucio slack channel might be a good place to confirm this behaviour. Can you please confirm it with the experts before we take any decision here?

Reply:
First, let me apologize for the longish comment I am about to make, but I want to clear any misunderstanding.  
Hi @nsmith- This should be the correct final result, Yes. But it does not happen like that. Even if a single `did` violates the `unique` constrain the whole batch is dropped. This I demonstrate few lines bellow. While @amaltaro, I think `Ricio Client` behaves exactly the opposite of what you say - it for sure tries to deal with the whole list in a single transaction, and that's exactly why it fails the hole list, but not only the ones who are the actual intruders.  

Let me paste some small snippet of code here [1].  And the output from it [2]. While creating the list of `dids`, I chose two datasets that already have previously created rules for them, and one dataset for which I am sure there is none. As you can see, at the end, the whole transaction is dropped, the `DuplicateRule` Exception is thrown for the whole call and the last dataset is left behind, without any notice that it is actually fine to have a rule created for it. So the caller of the method just skips the whole run presuming there are already rules created for his datasets (may be from previous runs or what so ever) and continues with his life, and there is the bug - the last dataset is left without a rule and unnoticed and no rule creation is repeated - ergo 'lost'.



[1]
```
In [35]: from rucio.client import Client
    ...: try:
    ...:     account = 'wma_test'
    ...:     rcl = Client(account=account)
    ...:     scope = 'cms'
    ...:     dids=[{'name': u'/RelValH125GGgluonfusion_13/CMSSW_8_1_0-RecoFullPU_2017PU_TaskChain_PUMCRecyc_Agent114_Validation_TEST_Alan_v18-v11/GEN-SIM-RECO', 'scope': scope},
    ...:           {'name': u'/RelValH125GGgluonfusion_13/CMSSW_8_1_0-DigiFullPU_2017PU_TaskChain_PUMCRecyc_Agent114_Validation_TEST_Alan_v18-v11/GEN-SIM-DIGI-RAW', 'scope': scope},
    ...:           {'name': u'/RelValH125GGgluonfusion_13/CMSSW_8_1_0-RecoFullPU_2017PU_TaskChain_PUMCRecyc_HG1803_Validation_TEST_Alan_v15-v11/GEN-SIM-RECO', 'scope': scope}]
    ...:     rseExpression = '(tier=2|tier=1)&cms_type=real&rse_type=DISK'
    ...:     copies = 1
    ...:     rcl.add_replication_rule(dids, copies, rseExpression)
    ...: except DuplicateRule as ex:
    ...:     msg = "\nA duplicate rule for: \naccount: %s \ndids: %s \nrseExpression: %s\ncopies: %s\n"
    ...:     ruleHistory = []
    ...:     for did in dids:
    ...:         didHistory = {}
    ...:         didHistory['did'] = did['name']
    ...:         didHistory['did_hist'] = []
    ...:         for hist in rcl.list_replication_rule_full_history(scope, did['name']):
    ...:             didHistory['did_hist'].append(hist)
    ...:         ruleHistory.append(didHistory)
    ...:     msg += "Rule History: \n%s\n"
    ...:     logger.debug(msg,
    ...:                  account,
    ...:                  pformat(dids),
    ...:                  rseExpression,
    ...:                  copies,
    ...:                  pformat(ruleHistory))
```


[2]
```
DEBUG:rucio.client.baseclient:no auth_type passed. Trying to get it from the environment variable RUCIO_AUTH_TYPE and config file.
DEBUG:rucio.client.baseclient:no creds passed. Trying to get it from the config file.
DEBUG:rucio.client.baseclient:no ca_cert passed. Trying to get it from the config file.
DEBUG:rucio.client.baseclient:got token from file
DEBUG:dogpile.lock:value creation lock <dogpile.cache.region._LockWrapper object at 0x7f64801ffd90> acquired
DEBUG:dogpile.lock:Calling creation function for previously expired value
DEBUG:dogpile.lock:Released creation lock
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): cmsrucio-int.cern.ch:80
DEBUG:urllib3.connectionpool:http://cmsrucio-int.cern.ch:80 "POST /rules/ HTTP/1.1" 409 None
DEBUG:urllib3.connectionpool:http://cmsrucio-int.cern.ch:80 "GET /rules/cms/%2FRelValH125GGgluonfusion_13%2FCMSSW_8_1_0-RecoFullPU_2017PU_TaskChain_PUMCRecyc_Agent114_Validation_TEST_Alan_v18-v11%2FGEN-SIM-RECO/history HTTP/1.1" 200 None
DEBUG:urllib3.connectionpool:http://cmsrucio-int.cern.ch:80 "GET /rules/cms/%2FRelValH125GGgluonfusion_13%2FCMSSW_8_1_0-DigiFullPU_2017PU_TaskChain_PUMCRecyc_Agent114_Validation_TEST_Alan_v18-v11%2FGEN-SIM-DIGI-RAW/history HTTP/1.1" 200 None
DEBUG:urllib3.connectionpool:http://cmsrucio-int.cern.ch:80 "GET /rules/cms/%2FRelValH125GGgluonfusion_13%2FCMSSW_8_1_0-RecoFullPU_2017PU_TaskChain_PUMCRecyc_HG1803_Validation_TEST_Alan_v15-v11%2FGEN-SIM-RECO/history HTTP/1.1" 200 0
DEBUG:__main__:
A duplicate rule for: 
account: wma_test 
dids: [{'name': u'/RelValH125GGgluonfusion_13/CMSSW_8_1_0-RecoFullPU_2017PU_TaskChain_PUMCRecyc_Agent114_Validation_TEST_Alan_v18-v11/GEN-SIM-RECO',
  'scope': 'cms'},
 {'name': u'/RelValH125GGgluonfusion_13/CMSSW_8_1_0-DigiFullPU_2017PU_TaskChain_PUMCRecyc_Agent114_Validation_TEST_Alan_v18-v11/GEN-SIM-DIGI-RAW',
  'scope': 'cms'},
 {'name': u'/RelValH125GGgluonfusion_13/CMSSW_8_1_0-RecoFullPU_2017PU_TaskChain_PUMCRecyc_HG1803_Validation_TEST_Alan_v15-v11/GEN-SIM-RECO',
  'scope': 'cms'}] 
rseExpression: (tier=2|tier=1)&cms_type=real&rse_type=DISK
copies: 1
Rule History: 
[{'did': u'/RelValH125GGgluonfusion_13/CMSSW_8_1_0-RecoFullPU_2017PU_TaskChain_PUMCRecyc_Agent114_Validation_TEST_Alan_v18-v11/GEN-SIM-RECO',
  'did_hist': [{u'account': u'wma_test',
                u'created_at': datetime.datetime(2020, 6, 30, 1, 34, 51),
                u'locks_ok_cnt': 9,
                u'locks_replicating_cnt': 0,
                u'locks_stuck_cnt': 0,
                u'rse_expression': u'(tier=2|tier=1)&cms_type=real&rse_type=DISK',
                u'rule_id': u'1f0ab297e4b54e1abf7c086ac012b9e9',
                u'state': u'OK',
                u'updated_at': datetime.datetime(2020, 6, 30, 1, 34, 51)},
               {u'account': u'wma_test',
                u'created_at': datetime.datetime(2020, 6, 30, 1, 36, 37),
                u'locks_ok_cnt': 9,
                u'locks_replicating_cnt': 0,
                u'locks_stuck_cnt': 0,
                u'rse_expression': u'(pnn=T2_CH_CERN|pnn=T1_US_FNAL_Disk)&cms_type=real&rse_type=DISK',
                u'rule_id': u'd1c29da28d6a409daf477e5934f754ce',
                u'state': u'OK',
                u'updated_at': datetime.datetime(2020, 6, 30, 1, 36, 37)}]},
 {'did': u'/RelValH125GGgluonfusion_13/CMSSW_8_1_0-DigiFullPU_2017PU_TaskChain_PUMCRecyc_Agent114_Validation_TEST_Alan_v18-v11/GEN-SIM-DIGI-RAW',
  'did_hist': [{u'account': u'wma_test',
                u'created_at': datetime.datetime(2020, 6, 30, 1, 34, 51),
                u'locks_ok_cnt': 18,
                u'locks_replicating_cnt': 0,
                u'locks_stuck_cnt': 0,
                u'rse_expression': u'(tier=2|tier=1)&cms_type=real&rse_type=DISK',
                u'rule_id': u'00e40fca61b5461e88b7bc235a47f531',
                u'state': u'OK',
                u'updated_at': datetime.datetime(2020, 6, 30, 1, 34, 52)},
               {u'account': u'wma_test',
                u'created_at': datetime.datetime(2020, 6, 30, 1, 36, 37),
                u'locks_ok_cnt': 18,
                u'locks_replicating_cnt': 0,
                u'locks_stuck_cnt': 0,
                u'rse_expression': u'(pnn=T2_CH_CERN|pnn=T1_US_FNAL_Disk)&cms_type=real&rse_type=DISK',
                u'rule_id': u'718bf04af115451f868f67c6ca71a1a5',
                u'state': u'OK',
                u'updated_at': datetime.datetime(2020, 6, 30, 1, 36, 37)}]},
 {'did': u'/RelValH125GGgluonfusion_13/CMSSW_8_1_0-RecoFullPU_2017PU_TaskChain_PUMCRecyc_HG1803_Validation_TEST_Alan_v15-v11/GEN-SIM-RECO',
  'did_hist': []}]
```

Reply:
I see. This is definitely a bug on the Rucio side. Would you mind creating a GH issue for them under: https://github.com/rucio/rucio ?

Regarding our real use case, the only scenario that I can think of where we would have some dids with rules, and some not (within the same rule uniqueness), is that a partial rule gets created for a sub-set of the dids that we provide. This can only happen in the following situations:
* if the client makes one HTTP call for each did in the list of dids (thus, single transaction on the server side)
* if the server breaks the request into multiple database transactions, one for each did

As Rucio relies heavily on Oracle, I'd say they probably implemented all the dids rule creation in a single transaction. Still, we should contact them to make sure that we only implemented what is needed. Please check their slack channel today during CERN time: https://github.com/rucio/rucio#getting-support

Reply:
Why not loop through the DIDs on the client side for safety?
You can catch the duplicate rule exception there, at the expense of more API calls.

Reply:
Hi @nsmith- , Yes we are doing exactly that, and it works. It is simply slower doing it on the client side, because it waits for getting the full history data per `did` and then it (eventually) parses it. This could be much faster doing it on the server side. And also it may be more clear to the other users too, if they could have a response from the server giving the information for which of the `dids` a rule was created and for which not.

Reply:
@amaltaro The question was forwarded to Rucio Support Team by mail.

Reply:
thanks, I suggest we leave it open for historical reference!